pkg/controller/releasemanager/releasemanager_controller.go

```diff
@@ -0,0 +1,407 @@
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	cluster := &picchuv1alpha1.Cluster{}
+	key := client.ObjectKey{request.Namespace, instance.Spec.Cluster}
+	if err := r.client.Get(context.TODO(), key, cluster); err != nil {
+		return reconcile.Result{}, err
+	}
+	incarnationList := &picchuv1alpha1.IncarnationList{}
+	labelSelector := client.MatchingLabels(map[string]string{
+		"medium.build/app":     instance.Spec.App,
+		"medium.build/cluster": instance.Spec.Cluster,
+	})
+	if err := r.client.List(context.TODO(), labelSelector, incarnationList); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	remoteClient, err := utils.RemoteClient(r.client, cluster)
+	if err != nil {
+		reqLogger.Error(err, "Failed to create remote client", "Cluster.Key", key)
+		return reconcile.Result{}, err
+	}
+	reqLogger.Info("Created remote client", "Cluster.Key", key)
+	manager := ReleaseManager{
+		Instance:        instance,
+		IncarnationList: incarnationList,
+		Cluster:         cluster,
+		Client:          remoteClient,
+		PicchuClient:    r.client,
+	}
+	if err := manager.SyncNamespace(); err != nil {
+		return reconcile.Result{}, err
+	}
+	if err := manager.SyncService(); err != nil {
+		return reconcile.Result{}, err
+	}
+	if err := manager.SyncDestinationRule(); err != nil {
+		return reconcile.Result{}, err
+	}
+	if err := manager.SyncVirtualService(); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+type ReleaseManager struct {
+	Instance        *picchuv1alpha1.ReleaseManager
+	IncarnationList *picchuv1alpha1.IncarnationList
+	Cluster         *picchuv1alpha1.Cluster
+	Client          client.Client
+	PicchuClient    client.Client
+}
+
+func (r *ReleaseManager) SyncNamespace() error {
+	labels := map[string]string{
+		"istio-injection": "enabled",
+	}
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   r.Instance.Spec.App,
+			Labels: labels,
+		},
+	}
+
+	op, err := controllerutil.CreateOrUpdate(context.TODO(), r.Client, namespace, func(runtime.Object) error {
+		namespace.ObjectMeta.SetLabels(labels)
+		return nil
+	})
+	log.Info("Namespace sync'd", "Op", op)
+	return err
+}
+
+func (r *ReleaseManager) SyncService() error {
+	portMap := map[string]corev1.ServicePort{}
+	for _, incarnation := range r.IncarnationList.Items {
+		for _, port := range incarnation.Spec.Ports {
+			_, ok := portMap[port.Name]
+			if !ok {
+				portMap[port.Name] = corev1.ServicePort{
+					Name:       port.Name,
+					Protocol:   port.Protocol,
+					Port:       port.Port,
+					TargetPort: intstr.FromString(port.Name),
+				}
+			}
+		}
+	}
+	ports := make([]corev1.ServicePort, len(portMap))
+	i := 0
+	for _, port := range portMap {
+		ports[i] = port
```

Comment:
```suggestion
		ports = append(ports, port)
```

Reply:
i read this is faster for getting values from a map, haven't tested though